lib/experimental/class-wp-theme-json-resolver-gutenberg.php

```diff
@@ -182,9 +182,9 @@ public static function get_merged_data( $origin = 'custom' ) {
 		}
 
 		$result = new WP_Theme_JSON_Gutenberg();
-		$result->merge( static::get_core_data() );
-		$result->merge( static::get_block_data() );
-		$result->merge( static::get_theme_data() );
+		$result->merge( apply_filters( 'global_styles_core', static::get_core_data() ) );
+		$result->merge( apply_filters( 'global_styles_blocks', static::get_block_data() ) );
+		$result->merge( apply_filters( 'global_styles_theme', static::get_theme_data() ) );
 		if ( 'custom' === $origin ) {
 			$result->merge( static::get_user_data() );
```

Comment:
```suggestion
			$result->merge( apply_filters( 'global_styles_user', static::get_user_data() ) );
```

Reply:
I'm not sure about this one. Is there any precedent in core for filtering user data? In principle I'd think that user data shouldn't be modified by anyone.

Reply:
The reason I'm suggesting this one, is that it's at the "end of the line" - where we'd expect to see a filter. Most WP filters are at the end of functions and filter the return values - including user data.
Ideally I'd like to see a `global_styles` filter at the `return` of the function as that would provide the maximum amount of flexibility without the need for separate filters per-context.
But if we do add separate filters for each item, then user styles will override any filters added - which would "neutralize" them. 
Perhaps we could add a `global_styles` filter at the return function, and pass the core/blocks/theme/user data as params to the filter?

Reply:
For user data: according to [how the API is evolving](https://github.com/WordPress/gutenberg/pull/44015#issuecomment-1241899708), adding a `global_styles_user` hook is a no-brainer. Though why someone would want to filter user styles and presets?

For the merged data: if we do this it'll be more involved for consumers to filter. At this point we've already merged the data coming from different origins (defaults, theme, and user), so consumers would need to have some understanding of what this means — for example, how we store the palettes for different origins. I'm less convinced about this one. It requires less knowledge from consumers if they just filter the bits they want to modify.

Reply:
I don't think user styles should be filterable. The use-case I can think of for that filter would be as a fail-safe for users making style/preset changes that a design system might not be allowing. However, that's probably something that should be controllable via `supports` properties?